File: class25/java/TCC2501-GeradorCodigo.java
// TCC2501-GeradorCodigo.java | Aula 25
// Gera código de três endereços (TAC) e código C a partir de uma AST.
// Espelho Java do TCC2501-GeradorCodigo.py (mesma AST e saídas).
import java.util.*;

public class TCC2501GeradorCodigo {

    // ---------- AST ----------
    static abstract class No {}
    static class Num extends No { double valor; Num(double v) { valor = v; } }
    static class Var extends No { String nome; Var(String n) { nome = n; } }
    static class BinOp extends No { String op; No esq, dir; BinOp(String o, No e, No d) { op = o; esq = e; dir = d; } }
    static class Assign extends No { String nome; No expr; Assign(String n, No e) { nome = n; expr = e; } }
    static class If extends No { No cond; List<No> thenBody, elseBody; If(No c, List<No> t, List<No> e) { cond = c; thenBody = t; elseBody = e; } }

    // ---------- Gerador TAC ----------
    static class GeradorTAC {
        List<String> codigo = new ArrayList<>();
        int tempCount = 0, labelCount = 0;

        String novoTemp() { return "t" + (++tempCount); }
        String novaLabel() { return "L" + (++labelCount); }

        String gerar(No no) {
            if (no instanceof Num) return String.valueOf(((Num) no).valor);
            if (no instanceof Var) return ((Var) no).nome;
            if (no instanceof BinOp) {
                BinOp b = (BinOp) no;
                String esq = gerar(b.esq), dir = gerar(b.dir);
                String t = novoTemp();
                codigo.add(t + " = " + esq + " " + b.op + " " + dir);
                return t;
            }
            if (no instanceof Assign) {
                Assign a = (Assign) no;
                String val = gerar(a.expr);
                codigo.add(a.nome + " = " + val);
                return a.nome;
            }
            if (no instanceof If) {
                If i = (If) no;
                String cond = gerar(i.cond);
                String lElse = novaLabel(), lFim = novaLabel();
                codigo.add("if " + cond + " == 0 goto " + lElse);
                for (No s : i.thenBody) gerar(s);
                codigo.add("goto " + lFim);
                codigo.add(lElse + ":");
                for (No s : i.elseBody) gerar(s);
                codigo.add(lFim + ":");
                return "";
            }
            throw new RuntimeException("Nó não suportado: " + no.getClass().getSimpleName());
        }
    }

    // ---------- Gerador C ----------
    static class GeradorC {
        List<String> linhas = new ArrayList<>();
        int indent = 0;
        String ind() { return "    ".repeat(indent); }

        String gerar(No no) {
            if (no instanceof Num) return String.valueOf(((Num) no).valor);
            if (no instanceof Var) return ((Var) no).nome;
            if (no instanceof BinOp) {
                BinOp b = (BinOp) no;
                return "(" + gerar(b.esq) + " " + b.op + " " + gerar(b.dir) + ")";
            }
            if (no instanceof Assign) {
                Assign a = (Assign) no;
                return ind() + a.nome + " = " + gerar(a.expr) + ";";
            }
            if (no instanceof If) {
                If i = (If) no;
                StringBuilder sb = new StringBuilder();
                sb.append(ind()).append("if (").append(gerar(i.cond)).append(") {\n");
                indent++;
                for (No s : i.thenBody) sb.append(gerar(s)).append("\n");
                indent--;
                sb.append(ind()).append("} else {\n");
                indent++;
                for (No s : i.elseBody) sb.append(gerar(s)).append("\n");
                indent--;
                sb.append(ind()).append("}");
                return sb.toString();
            }
            throw new RuntimeException("Nó não suportado: " + no.getClass().getSimpleName());
        }
    }

    public static void main(String[] args) {
        // AST para: if (x > 0) y = x * 2; else y = 0;
        If ast = new If(
            new BinOp(">", new Var("x"), new Num(0)),
            Arrays.asList(new Assign("y", new BinOp("*", new Var("x"), new Num(2)))),
            Arrays.asList(new Assign("y", new Num(0)))
        );

        System.out.println("=== Código de Três Endereços (TAC) ===");
        GeradorTAC genTac = new GeradorTAC();
        genTac.gerar(ast);
        for (String linha : genTac.codigo) System.out.println("  " + linha);

        System.out.println("\n=== Código C Gerado ===");
        GeradorC genC = new GeradorC();
        System.out.println(genC.gerar(ast));
    }
}

File: class25/python/TCC2501-GeradorCodigo.py
#!/usr/bin/env python3
"""TCC2501-GeradorCodigo.py | Aula 25
Gera código de três endereços (TAC) e código C a partir de uma AST.
"""
from dataclasses import dataclass
from typing import List


@dataclass
class No:
    pass


@dataclass
class Num(No):
    valor: float


@dataclass
class Var(No):
    nome: str


@dataclass
class BinOp(No):
    op: str
    esq: No
    dir: No


@dataclass
class Assign(No):
    nome: str
    expr: No


@dataclass
class If(No):
    cond: No
    then_body: List[No]
    else_body: List[No]


class GeradorTAC:
    """Gera código de três endereços (TAC) a partir da AST."""
    def __init__(self):
        self.codigo: List[str] = []
        self.temp_count = 0
        self.label_count = 0

    def novo_temp(self) -> str:
        self.temp_count += 1
        return f"t{self.temp_count}"

    def nova_label(self) -> str:
        self.label_count += 1
        return f"L{self.label_count}"

    def gerar(self, no: No) -> str:
        if isinstance(no, Num):
            return str(no.valor)
        if isinstance(no, Var):
            return no.nome
        if isinstance(no, BinOp):
            esq = self.gerar(no.esq)
            dir_ = self.gerar(no.dir)
            t = self.novo_temp()
            self.codigo.append(f"{t} = {esq} {no.op} {dir_}")
            return t
        if isinstance(no, Assign):
            val = self.gerar(no.expr)
            self.codigo.append(f"{no.nome} = {val}")
            return no.nome
        if isinstance(no, If):
            cond = self.gerar(no.cond)
            l_else = self.nova_label()
            l_fim = self.nova_label()
            self.codigo.append(f"if {cond} == 0 goto {l_else}")
            for s in no.then_body:
                self.gerar(s)
            self.codigo.append(f"goto {l_fim}")
            self.codigo.append(f"{l_else}:")
            for s in no.else_body:
                self.gerar(s)
            self.codigo.append(f"{l_fim}:")
            return ""
        raise ValueError(f"Nó não suportado: {type(no)}")


class GeradorC:
    """Gera código C a partir da AST."""
    def __init__(self):
        self.codigo: List[str] = []
        self.indent = 0

    def _ind(self) -> str:
        return "    " * self.indent

    def gerar(self, no: No) -> str:
        if isinstance(no, Num):
            return str(no.valor)
        if isinstance(no, Var):
            return no.nome
        if isinstance(no, BinOp):
            return f"({self.gerar(no.esq)} {no.op} {self.gerar(no.dir)})"
        if isinstance(no, Assign):
            return f"{self._ind()}{no.nome} = {self.gerar(no.expr)};"
        if isinstance(no, If):
            linhas = [f"{self._ind()}if ({self.gerar(no.cond)}) {{"]
            self.indent += 1
            for s in no.then_body:
                linhas.append(self.gerar(s))
            self.indent -= 1
            linhas.append(f"{self._ind()}}} else {{")
            self.indent += 1
            for s in no.else_body:
                linhas.append(self.gerar(s))
            self.indent -= 1
            linhas.append(f"{self._ind()}}}")
            return "\n".join(linhas)
        raise ValueError(f"Nó não suportado: {type(no)}")


if __name__ == "__main__":
    # AST para: if (x > 0) y = x * 2; else y = 0;
    ast = If(
        cond=BinOp(">", Var("x"), Num(0)),
        then_body=[Assign("y", BinOp("*", Var("x"), Num(2)))],
        else_body=[Assign("y", Num(0))],
    )

    print("=== Código de Três Endereços (TAC) ===")
    gen_tac = GeradorTAC()
    gen_tac.gerar(ast)
    for linha in gen_tac.codigo:
        print(f"  {linha}")

    print("\n=== Código C Gerado ===")
    gen_c = GeradorC()
    print(gen_c.gerar(ast))
